File: app/src/main/java/com/darelbitsy/dbweather/helper/provider/LocationSuggestionProvider.java
package com.darelbitsy.dbweather.helper.provider;

import android.app.SearchManager;
import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.provider.BaseColumns;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import com.darelbitsy.dbweather.controller.api.adapters.helper.GeoNamesHelper;
import com.darelbitsy.dbweather.helper.holder.ConstantHolder;
import com.darelbitsy.dbweather.model.geonames.GeoName;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Darel Bitsy on 04/04/17.
 * Location Suggestion provider
 */

public class LocationSuggestionProvider extends ContentProvider {
    private GeoNamesHelper mGeoNamesHelper;
    private final CompositeDisposable mCompositeDisposable = new CompositeDisposable();
    public static final List<GeoName> mListOfLocation = new ArrayList<>();

    @Override
    public boolean onCreate() {
        mGeoNamesHelper = GeoNamesHelper.newInstance(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        final String userQuery = uri.getLastPathSegment();

        final MatrixCursor matrixCursor = new MatrixCursor(new String[] {
                BaseColumns._ID,
                SearchManager.SUGGEST_COLUMN_TEXT_1,
                SearchManager.SUGGEST_COLUMN_TEXT_2
        }, mListOfLocation.size());

        if (userQuery != null && !userQuery.isEmpty()) {
            mCompositeDisposable.add(mGeoNamesHelper.getLocationFromApi(userQuery)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(new GetSuggestions()));

            for (int index = 0; index < mListOfLocation.size(); index ++) {
                GeoName location = mListOfLocation.get(index);

                matrixCursor.addRow(new Object[] {index, location.getName(), location.getCountryName()});
            }
        }

        return matrixCursor;
    }

    private class GetSuggestions extends DisposableSingleObserver<List<GeoName>> {
        @Override
        public void onSuccess(List<GeoName> geoNames) {
            mListOfLocation.clear();
            mListOfLocation.addAll(geoNames);
        }

        @Override
        public void onError(Throwable e) {
            Log.i(ConstantHolder.TAG, "Error in location provider: " + e.getMessage());
        }
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: app/src/main/java/com/dbeginc/dbweather/utils/contentprovider/LocationSuggestionProvider.kt
/*
 *  Copyright (C) 2017 Darel Bitsy
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

package com.dbeginc.dbweather.utils.contentprovider

import android.app.SearchManager
import android.content.ContentValues
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.provider.BaseColumns
import com.dbeginc.dbweather.R
import com.dbeginc.dbweather.utils.utility.WEATHER_SEARCH_RESULTS
import com.dbeginc.dbweatherdomain.Logger
import com.dbeginc.dbweatherdomain.ThreadProvider
import com.dbeginc.dbweatherdomain.repositories.WeatherRepository
import com.dbeginc.dbweatherweather.viewmodels.toUi
import dagger.Lazy
import dagger.android.DaggerContentProvider
import javax.inject.Inject

/**
 * Created by Darel Bitsy on 04/04/17.
 *
 * Location Suggestion provider
 */
class LocationSuggestionProvider : DaggerContentProvider() {
    @Inject lateinit var model: Lazy<WeatherRepository>
    @Inject lateinit var logger: Lazy<Logger>
    @Inject lateinit var threads: Lazy<ThreadProvider>


    override fun query(uri: Uri, projection: Array<String>?,
                       selection: String?, selectionArgs: Array<String>?,
                       sortOrder: String?): Cursor? {


        val userQuery = uri.lastPathSegment

        val matrixCursor = MatrixCursor(arrayOf(
                BaseColumns._ID,
                SearchManager.SUGGEST_COLUMN_ICON_1,
                SearchManager.SUGGEST_COLUMN_TEXT_1,
                SearchManager.SUGGEST_COLUMN_TEXT_2
        ), 3)

        if (userQuery != null && userQuery != SearchManager.SUGGEST_URI_PATH_QUERY && userQuery.isNotEmpty()) {
            model.get().getLocations(userQuery)
                    .map { locations -> locations.map { location -> location.toUi() } }
                    .observeOn(threads.get().UI)
                    .subscribe(WEATHER_SEARCH_RESULTS::onNext, logger.get()::logError)
        }

        WEATHER_SEARCH_RESULTS
                .value
                ?.forEachIndexed { index, location ->
                    matrixCursor.addRow(arrayOf(
                            index,
                            R.drawable.ic_city_location,
                            location.name,
                            location.countryName
                    ))
                }

        return matrixCursor
    }

    override fun getType(uri: Uri): String? = null

    override fun insert(uri: Uri, values: ContentValues?): Uri? = null

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int = 0

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int = 0

}